### PR TITLE
Init DAP protocol in IO plugin.

### PR DESCRIPTION
**Pull request by gogo2464**
- [ ] Mark this if you consider it ready to merge
- [ ] I've added tests (optional)
- [ ] I wrote some lines in the [book](https://github.com/radareorg/radare2book) (optional)

**Description**

Will solve: https://github.com/radareorg/radare2/issues/12128.

### CODE REVIEW

File: libr/io/p/io_dap.c [hunk 1]
```diff
@@ -0,0 +1,124 @@
...
+#include <errno.h>
+
+#include "io_dap.h"
+
+#define URI_PREFIX "foo://"
+
+extern RIOPlugin r_io_plugin_dap; // forward declaration
+
+static bool __plugin_open(RIO *io, const char *pathname, bool many) {
+    return (strncmp(pathname, URI_PREFIX, strlen(URI_PREFIX)) == 0);
```
Comment (trufae) [inline libr/io/p/io_dap.c:15]: use tabs

Comment (trufae) [inline libr/io/p/io_dap.c:15]: use r_str_startswith()

File: libr/io/p/io_dap.c [hunk 2]
```diff
@@ -0,0 +1,124 @@
...
+}
+
+static RIODesc *__open(RIO *io, const char *pathname, int flags, int mode) {
+    RIODesc *ret = NULL;
+    RIOFoo *rio_foo = NULL;
+
+    printf("%s\n", __func__);
```
Comment (trufae) [inline libr/io/p/io_dap.c:22]: remove debug printfs

File: libr/io/p/io_dap.c [hunk 3]
```diff
@@ -0,0 +1,124 @@
...
+
+    if (!__plugin_open(io, pathname, 0))
```
Comment (trufae) [inline libr/io/p/io_dap.c:24]: use tabs and space before (

File: libr/io/p/io_dap.c [hunk 4]
```diff
@@ -0,0 +1,124 @@
...
+        return ret;
+
+    return r_io_desc_new (io, &r_io_plugin_dap, pathname, flags, mode, rio_foo);
+}
+
+static int __close(RIODesc *fd) {
+    RIOFoo *rio_foo = NULL;
+
+    printf("%s\n", __func__);
```
Comment (trufae) [inline libr/io/p/io_dap.c:33]: remove log msg

File: libr/io/p/io_dap.c [hunk 5]
```diff
@@ -0,0 +1,124 @@
...
+
+    if (!__plugin_open(io, pathname, 0))
+        return ret;
+
+    return r_io_desc_new (io, &r_io_plugin_dap, pathname, flags, mode, rio_foo);
+}
+
+static int __close(RIODesc *fd) {
+    RIOFoo *rio_foo = NULL;
```
Comment (trufae) [inline libr/io/p/io_dap.c:31]: initialize and assign in the same line after the if

File: libr/io/p/io_dap.c [hunk 6]
```diff
@@ -0,0 +1,124 @@
...
+
+    printf("%s\n", __func__);
+    if (!fd || !fd->data)
```
Comment (trufae) [inline libr/io/p/io_dap.c:34]: missing braces

**Review by trufae — CHANGES_REQUESTED** [review]
supporting DAP will be a cool addition. you can see how other io-debug plugins work so you dont need to write a debug plugin to make it work as a backend. (just use the `e dbg.backend=io`)

**Comment by gogo2464** [comment]
I used this template: https://wenzel.github.io/2018/04/15/radare2-io-plugin-tutorial.html nothing is ready yet.

**Comment by trufae** [comment]
the error:
```
'/home/runner/work/radare2/radare2/libr/parse/p'
[1451](https://github.com/radareorg/radare2/actions/runs/3196044580/jobs/5217456357#step:7:1452)
tcc: error: file '/home/runner/work/radare2/radare2/libr/../shlr/dap/lib/libgdbr.a' not found
[1452](https://github.com/radareorg/radare2/actions/runs/3196044580/jobs/5217456357#step:7:1453)
make[3]: *** [/home/runner/work/radare2/radare2/libr/..//libr/rules.mk:91: libr_io.so] Error 1
```

the problem

```
LDFLAGS+=$(SHLR)/dap/lib/libgdbr.$(EXT_AR)
```

**Comment by gogo2464** [comment]
@trufae fixed thank you so much!

For the other issues, this is just a template from internet. We could import the doc on r2book and r2 blog from https://wenzel.github.io/2018/04/15/radare2-io-plugin-tutorial.html and then update it.

**Comment by trufae** [comment]
i cant repro the ptrace-attach problem you reported yesterday on the irc. the plugin is called and behaves as expected

**Comment by trufae** [comment]
the failing linux-nocs job in the ci is because you are indenting with spaces instead of tabs. please run `sys/lint.sh` before commiting so you'll find some good hints to follow the coding style

**Comment by gogo2464** [comment]
> i cant repro the ptrace-attach problem you reported yesterday on the irc. the plugin is called and behaves as expected

I think I had untracked changes. Now that  git clone my own PR again from github, all seems fine.

**Comment by gogo2464** [comment]
> i cant repro the ptrace-attach problem you reported yesterday on the irc. the plugin is called and behaves as expected

I also used remote debuging from emulicious to test. Did you do so?

File: libr/io/p/io_dap.c (hunk 2 above)
Comment (gogo2464) [reply libr/io/p/io_dap.c:37]: could I remove it in the future?

**Comment by trufae** [comment]
Closing this PR, after 2 years there is no relevant code in here, over time it got conflicts, so i think its better to close it

**Comment by gogo2464** [comment]
Yes sorry. I am very sorry :(

**Pull request closed without merging**